trace-dispatcher/doc/trace-dispatcher.md

```diff
@@ -0,0 +1,798 @@
+   5. [Documentation generation](#Documentation-generation)
+5. [Appendix](#Appendix)
+   1. [Decisions](#Decisions)
+   2. [Future work](#Future-work)
+
+# Document status
+
+Work in progress.
+
+To do list:
+
+* [x] Finish editing.
+* [x] [Decide inline trace type annotation with trace function](#Decide-inline-trace-type-annotation-with-trace-function)
+* [x] [Decide tracer definedness](#Decide-tracer-definedness)
+* [x] [Decide tracer name definition](#Decide-tracer-name-definition)
+* [x] [Decide inline trace type annotation with trace function 2](#Decide-inline-trace-type-annotation-with-trace-function-2)
+* [x] [Decide on explicit trace filtering](#Decide-on-explicit-trace-filtering)
+* [x] [Decide on privately combinator](#Decide-on-privately-combinator)
+* [x] [Decide extra privacy tagging](#Decide-extra-privacy-tagging)
+* [x] [Decide on type errors instead of silent dropping of messages](#Decide-on-type-errors-instead-of-silent-dropping-of-messages)
+* [ ] [Decide on more direct interface to EKG metrics](#Decide-on-more-direct-interface-to-ekg-metrics)
+* [x] [Decide on dispatcher detail level control](#Decide-on-dispatcher-detail-level-control)
+* [x] [Decide namespace-aware configuration](#Decide-namespace-aware-configuration)
+* [ ] [Discuss impact of missing documentation entries](#Discuss-impact-of-missing-documentation-entries)
+* [x] [Decide missing configuration](#Decide-missing-configuration)
+* [x] [Decide complete configuration](#Decide-complete-configuration)
+* [ ] [Discuss possibility of pure, thread-safe aggregation](#Discuss-possibility-of-pure-thread-safe-aggregation)
+* [ ] [Decide trace-outs types](#Decide-trace-outs-types)
+* [ ] Agree on editing.
+* [ ] Final proofreading.
+* [ ] Feedback.
+
+# Introduction
+
+## Motivation
+
+`trace-dispatcher` is an attempt to enable design and implementation of simple, efficient __tracing systems__, one that has a reduced footprint in the executed program, has a more pleasant API, and provides self-documenting features.
+
+## Design decisions
+
+Key design decisions were:
+
+1. Retaining the separation of concerns in the frontend side, as provided by the `contra-tracer` library.  The client code should not need to concern itself with any details beyond passing the traces down to the system.
+2. Rely on __trace combinators__ primarily, as opposed to opting for a typeclass heavy API.
+3. Separation of data plane and control plane:  high-frequency events incur minimal processing on the data-plane, whereas complicated configuration logic only happens on the control plane, and that is proportional to infrequent reconfiguration events.
+4. A tougher stance on separation of concerns in the backend side:  we choose to move expensive trace processing to an external process.
+5. A measure of backward compatibility with the previous logging system.
+6. Retaining the global namespace for all traces.
```

Comment:
I think "the" should be "a" here? You're saying that all traces are keyed on some global key type?

Reply:
Yes, all traces are keyed by a list of strings/text.